index.d.ts

```diff
@@ -788,10 +767,207 @@ type ExecaCommonError = {
 	This is `undefined` unless the subprocess exited due to an `error` event or a timeout.
 	*/
 	originalMessage?: string;
+
+	/**
+	Underlying error, if there is one. For example, this is set by `.kill(error)`.
+
+	This is usually an `Error` instance.
+	*/
+	cause?: unknown;
+
+	/**
+	The name of the signal (like `SIGTERM`) that terminated the subprocess, sent by either:
+	- The current process.
+	- Another process. This case is [not supported on Windows](https://nodejs.org/api/process.html#signal-events).
+
+	If a signal terminated the subprocess, this property is defined and included in the error message. Otherwise it is `undefined`.
+	*/
+	signal?: string;
+
+	/**
+	A human-friendly description of the signal that was used to terminate the subprocess. For example, `Floating point arithmetic error`.
+
+	If a signal terminated the subprocess, this property is defined and included in the error message. Otherwise it is `undefined`. It is also `undefined` when the signal is very uncommon which should seldomly happen.
+	*/
+	signalDescription?: string;
+
+	/**
+	Node.js-specific [error code](https://nodejs.org/api/errors.html#errorcode), when available.
+	*/
+	code?: string;
 } & Error;
 
-export type ExecaError<OptionsType extends Options = Options> = ExecaCommonResult<false, OptionsType> & ExecaCommonError;
-export type ExecaSyncError<OptionsType extends SyncOptions = SyncOptions> = ExecaCommonResult<true, OptionsType> & ExecaCommonError;
+declare class CommonErrorClass<IsSync extends boolean, OptionsType extends Options = Options> extends Error {
```

Comment:
This copies all fields and their descriptions from `CommonError` and `CommonResult`.

This is because of TypeScript shortcomings:
  - This must be a `class` so that `instanceof` and `InstanceType` work
  - Mixin and multiple inheritance are not possible in an ambient file
  - Classes cannot use `&` or merge types from another type, except through `extends` or composition, which cannot be used here.

Avoiding this duplication would be great. If you have any idea on how to prevent this, please let me know.

Reply:
I've hit this issue in the past too and not sure there is a better solution. It would be useful with a short code comment that points out why we are doing it like this, because if you are not aware of the limitation, the code looks unnecessarily boilerplaty. Maybe just add a URL to the above comment.

Reply:
I've found a solution. Having a common class for both success and failed results, from which they both inherit.
Common result's type type error properties as `string | undefined`
Success result's type narrows error properties to `never`.
Failed result's type narrows  error properties to `string` (no `undefined`).

I've updated the PR with this.
No more duplication. :tada: